### PR TITLE
feat(user): hide softcore progress on profiles if none exists

### PR DESCRIPTION
**Pull request by wescopeland**
This PR adjusts how stats are rendered on user profiles.

Currently on prod, this is entirely template-driven. This PR adds a new component, _arranged-stat-items.blade.php_, which accepts a single prop: `stats`:

```html
<x-user.profile.arranged-stat-items
    :stats="[
        $developerStats['setsWorkedOnStat'],                    $developerStats['codeNotesCreatedStat'],
        $developerStats['achievementsUnlockedByPlayersStat'],   $developerStats['leaderboardsCreatedStat'],
        $developerStats['pointsAwardedToPlayersStat'],          $developerStats['openTicketsStat']
    ]"
/>
```

With this prop, it's easy from a DX standpoint to see how the stats will be rendered on desktop. For mobile, the columns are stacked on top of each other to keep data presentation cohesive (just like on prod today).

The controller is now responsible for building these stat entities, and the templates can arrange them however and whenever they want.

_player-stats.blade.php_ now picks and chooses what stats to render based on if the user is hardcore-only, softcore-only, or mixed. If the player has no progress from one of the modes, they don't see stats for that mode.

To conserve space, if a user is not a mixed-progress player, points and rank are not shown in player stats as they already exist in the primary meta area near the top of the page.

**Hardcore-only player stats**
Screenshot 2024-02-03 at 12 48 01 PM

---

Additional functionality changes for stats:
* A requested stat has been added for hardcore users: "Total games beaten". This pulls directly from the `PlayerStat` model used for the beaten games leaderboard.
* "Average completion percentage" has been renamed to "Average completion".
* "Started games beaten" now ignores subsets.

### CODE REVIEW

File: app/Community/Components/UserProfileMeta.php [hunk 1]
```diff
@@ -28,58 +29,255 @@ public function __construct(
...
+        // Hardcore site rank
+        $hardcoreSiteRankValue = '';
+        if ($userMassData['Untracked']) {
+            $hardcoreSiteRankValue = 'Untracked';
+        } elseif ($hardcorePoints > 0 && !$hardcoreRankMeta['rank']) {
+            $hardcoreSiteRankValue = 'requires ' . Rank::MIN_POINTS . ' points';
+        } elseif ($hardcorePoints === 0 && !$hardcoreRankMeta['rank']) {
+            $hardcoreSiteRankValue = 'none';
+        } else {
+            $hardcoreSiteRankValue = "#" . localized_number($hardcoreRankMeta['rank']) . " of " . localized_number($hardcoreRankMeta['numRankedUsers']);
```
Comment (Jamiras) [inline app/Community/Components/UserProfileMeta.php:160]: I imagine it's hard to do with the new construction paradigm, but I prefer just having the rank linked.
image
vs
image

It seems like you've got something working - White Points appear differently hand normal hardcore points.

File: app/Community/Components/UserProfileMeta.php [hunk 2]
```diff
@@ -28,58 +29,255 @@ public function __construct(
...
+
+        return array_merge(
+            compact(
+                'averageCompletionStat',
+                'averagePointsPerWeekStat',
+                'pointsLast30DaysStat',
+                'pointsLast7DaysStat',
+                'startedGamesBeatenPercentageStat',
+                'totalGamesBeatenStat',
+            ),
```
Comment (Jamiras) [inline app/Community/Components/UserProfileMeta.php:269]: Suggest reordering these so the beaten/completion stuff is in one column, and the points stuff is in the other.

Current:
image

Proposed:
image

This also flows better in mobile.
Current:
image

Proposed:
image

File: resources/views/community/components/user/profile/player-stats.blade.php [hunk 3]
```diff
@@ -32,71 +24,70 @@ class="relative w-full px-2 pt-2 bg-embed rounded mb-6 pb-4 transition-all"
...
+        <x-user.profile.arranged-stat-items
+            :stats="[
+                $playerStats['hardcoreAchievementsUnlockedStat'],   $playerStats['retroRatioStat'],
+                $playerStats['totalGamesBeatenStat'],               $playerStats['startedGamesBeatenPercentageStat'],   
+            ]"
+        />
+    @elseif ($primaryMode === 'softcore')
+        <x-user.profile.arranged-stat-items
+            :stats="[
+                $playerStats['softcoreAchievementsUnlockedStat'],   $playerStats['startedGamesBeatenPercentageStat'],
```
Comment (Jamiras) [inline resources/views/community/components/user/profile/player-stats.blade.php:57]: I assume it's intentional to keep things in matching rows, but Total Games Beaten doesn't show for pure-softcore players.

File: app/Community/Components/UserProfileMeta.php (hunk 1 above)
Comment (wescopeland) [reply app/Community/Components/UserProfileMeta.php]: Addressed in latest.

File: app/Community/Components/UserProfileMeta.php (hunk 2 above)
Comment (wescopeland) [reply app/Community/Components/UserProfileMeta.php:272]: No problem, I've changed the sort order for both the mixed and non-mixed "see more" content in latest.

File: resources/views/community/components/user/profile/player-stats.blade.php (hunk 3 above)
Comment (wescopeland) [reply resources/views/community/components/user/profile/player-stats.blade.php:57]: Right now it's coming purely from `PlayerStat`, which drives the beaten game leaderboard and only takes hardcore numbers into account.

```php
// Total games beaten
$totalGamesBeaten = (int) PlayerStat::where('user_id', $user->ID) // keep in mind, this is hardcore only.
    ->where('system_id', null)
    ->sum('value');
$totalGamesBeatenStat = [
    'label' => 'Total games beaten',
    'value' => localized_number($totalGamesBeaten),
    'isMuted' => !$totalGamesBeaten,
    // TODO: href to 'ranking.beaten-games' for a specific user.
];
```

**Review by Jamiras — APPROVED** [review]

**Pull request merged**